Add .md to README and clarify what example does

mjcross opened the pull request:
1. For pico_w/wifi/mqtt, rename README to README.md to improve how the content renders in an IDE or when browsing the online repository.
2. Clarify that the example doesn't re-publish the temperature reading if it hasn't changed; so that users can understand why the updates can appear sporadic.
3. Document how the led status can be checked by subscribing to /led/status

Commit:
Add .md to README and clarify what example does

Renamed README to README.md to improve how it renders in an IDE and when
browsing the online repository. Improved the example description to
clarify why the temperature updates might appear unexpectedly sporadic
to the user; and document how to read the current led status.

Commit:
Improve comments in mqtt_client example

lurch commented on pico_w/wifi/mqtt/mqtt_client.c:
```diff
@@ -368,7 +370,13 @@ int main(void) {
+    // because it's a read-only operation
     while (!state.connect_done || mqtt_client_is_connected(state.mqtt_client_inst)) {
+        // As supplied the example configures cyw43_arch for thread_safe_background operation
+        // by linking `ico_cyw43_arch_lwip_threadsafe_background` in CMakeLists.txt, so the 
```
Typo here? Did you mean `pico_cyw43_arch_lwip_threadsafe_background` ?

peterharperuk commented:
Looks good. Just the one typo pointed out by @lurch

lurch commented:
FYI #498 is where this example was first added.

Commit:
fix typo in comment

mjcross commented:
Thanks @peterharperuk and well spotted @lurch - fixed in my [a23bca5](https://github.com/mjcross/pico-examples/commit/a23bca58ddf7b6de1f9e2d5e011b5e521c12d95b). One of these days I'll manage to get a PR through without a typo 8-)

Commit:
protect call to lwIP mqtt_client_new()

Note that start_client() may be called direct from main() - i.e. not in a callback

mjcross commented:
If the dns query returns immediately then start_client() is called directly from main(); so shouldn't the call to mqtt_client_new() be protected? The lwIP implementation has LWIP_ASSERT_CORE_LOCKED

mjcross commented:
With the greatest respect, although this example is very thorough it is also quite hard to follow. 
Might I suggest that In the spirit of the other examples it would be clearer to use global variables and move some of the work outside the callbacks?

peterharperuk commented:
> If the dns query returns immediately then start_client() is called directly from main(); so shouldn't the call to mqtt_client_new() be protected? The lwIP implementation has LWIP_ASSERT_CORE_LOCKED

Yes

peterharperuk commented:
If you have suggestions for other improvements and fixes. I would suggest raising them as separate pull requests.

lurch commented on pico_w/wifi/mqtt/mqtt_client.c:
```diff
@@ -368,7 +372,13 @@ int main(void) {
+        // As supplied the example configures cyw43_arch for thread_safe_background operation
+        // by linking `pico_cyw43_arch_lwip_threadsafe_background` in CMakeLists.txt, so the 
+        // following two lines are unnecessary (but do no harm). However you will need them 
+        // if you reconfigure the build to use cyw43_arch in polling mode.
```
If the lines are needed in one mode but not in a different mode, maybe wrapping them in an `#ifdef` would make that more obvious? (Which I think is a pattern I've seen used in other examples?)

mjcross replied:
Agreed - see [12bbe5c](https://github.com/raspberrypi/pico-examples/pull/712/commits/12bbe5c154cefef55d05e52896ff1e2cf6d211ef)

Commit:
only poll cyw43_arch() in main() if we are in polling mode

mjcross commented:
> > If the dns query returns immediately then start_client() is called directly from main(); so shouldn't the call to mqtt_client_new() be protected? The lwIP implementation has LWIP_ASSERT_CORE_LOCKED
> 
> Yes

The code as stands also calls mqtt_client_is_connected(), which the lwIP code also has the CORE_LOCKED assertion. Not sure whether that's safe or not?
